cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+
+#include <cub/cub.cuh>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {    // anonymous
+
+template <typename T, typename Filter, bool has_validity>
+__global__
+void copy_if_else_kernel(  column_device_view const lhs,
+                           column_device_view const rhs,
+                           Filter filter,
+                           mutable_column_device_view out,
+                           cudf::size_type * __restrict__ const null_count)
+{   
+   const cudf::size_type tid = threadIdx.x + blockIdx.x * blockDim.x;   
+   const int w_id = tid / warp_size;
+   // begin/end indices for the column data
+   cudf::size_type begin = 0;
+   cudf::size_type end = lhs.size();   
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via 
+   // __ballot_sync()
+   cudf::size_type w_begin = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(begin);
+   cudf::size_type w_end = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(end);
+
+   // lane id within the current warp
+   const int w_lane_id = threadIdx.x % warp_size;
+
+   // store a null count for each warp in the block
+   constexpr cudf::size_type b_max_warps = 32;
+   __shared__ uint32_t b_warp_null_count[b_max_warps];
```

Comment:
Just store the count in a register and then call `single_lane_block_sum_reduce` to reduce a bunch of extra code in this kernel. 

https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/include/cudf/detail/utilities/cuda.cuh#L82

You can see how I used it in `valid_if`. https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/include/cudf/detail/valid_if.cuh#L45

Reply:
This comment was redundant with Mark's earlier one :) https://github.com/rapidsai/cudf/pull/3331#discussion_r345527508